### DISCUSSION TITLE
Can GitHub actions directly edit files in a repository?

### DISCUSSION THREAD

**Discussion by jpadfield** [community/community#25234] [Actions]
<p>Hi,</p>
<p>“Can GitHub Actions run a PHP file that can read, delete and create files directly within a repository?”</p>
<p>I have a simple repository that builds a series of html pages, based on a few json files. The build process is carried out by a PHP file. The idea is to simplify the process of making a set of consistent GitHub web-pages for a given repository.</p>
<p>I have been running the PHP file on a local machine in a cloned copy of the repository: it removes the old html files in the docs folder, reads in the json files and then creates a new set of html pages into the local docs folder, based on the content of the json files.</p>
<p>The local copy of the repository can then be pushed back to GitHub and the GitHub pages for the project are updated.</p>
<p>What I was wondering is, can this process simply be run as an “action” within GitHub? Basically the plan would be if one of three named json files are updated, the action could run and automatically rebuild the html files.</p>
<p>I can see lots of example actions that run files but I can not see any that actually directly interact with the existing repository files, is it possible? I have seen some talking about automating push processes but I could not see how it could include the files created by the php file that is run within the action.</p>
<p>I would also be very interested if it was possible to call the php file from a different repository but still run it locally within your own repository, but this is less important. :slight_smile: (This would allow a single build process to be used across a small set of different repositories)</p>
<p>I am hoping that this, if possible should be reasonably simple.</p>
<p>Any suggestions most welcome.</p>
<p>Thanks</p>
<p>Joe</p>

**Comment by logankilpatrick** [comment]
<p>Hey <a class="mention" href="/jpadfield">@jpadfield</a>, you can most definitely have an action that edits internal files (except those in the “.github” folder. </p>
<p>I do the following in one of my repo’s which calls a python script that parses all my source code and updates a todo.txt file and then commits it to the repo. </p>
<pre><code>jobs:
  build:

    runs-on: macos-latest

    steps:
    - uses: actions/checkout@v1
    - name: Run a one-line script
      run: python ./.github/workflows/todo_List.py
    - name: switching from HTTPS to SSH
      run: git remote set-url origin ${{ secrets.ssh }}
    - name: check for changes
      run: git status
    - name: stage changed files
      run: git add .
    - name: commit changed files
      run: git commit -m "Auto updating TODO.txt"
    - name: fetch from master
      run: git fetch origin master
    - name: push code to master
      run: git push origin HEAD:master
</code></pre>
<p>Let me know if you have more questions I can help with! </p>

**Reply by shner-elmo** [reply]
Would you be able to link the repo you have this script in? Thanks.

**Reply by mgilardi** [reply]
@shner-elmo Looks like its this one: https://github.com/logankilpatrick/TODO-List-Updater

**Reply by shner-elmo** [reply]
Late reply, but thanks a lot

**Comment by weide-zhou** [comment]
<p>Hi <a class="mention" href="/jpadfield">@jpadfield</a> ,</p>
<p>Yes, github actions can achieve this. </p>
<p>Step1: You need the get the source code with actions/checkout, if the code exists on other repo, you need to add parameters ‘repository:’ and ‘token’, please check the <a href="https://github.com/actions/checkout#usage" rel="nofollow noopener">usage</a> accordingly.</p>
<p>Step2: You need to rebuild the html files.</p>
<p>Step3. You can push the code to repo with actions ‘<a href="https://github.com/ad-m/github-push-action" rel="nofollow noopener">ad-m/github-push-action</a>’, if you want to push to another repo, please add ‘repository’ and ‘github_token’ parameter for the action. </p>
<p>Here is my <a href="https://github.community/t/possible-to-commit-files-after-workflow-runs/17824">answer</a> in a similar ticket which push back the code to repository, hope it helps!</p>

**Comment by jpadfield** [comment]
<p>Hi <a class="mention" href="/weide-zhou">@weide-zhou</a> &amp; <a class="mention" href="/logankilpatrick">@logankilpatrick</a> </p>
<p>THank you for your responses it is great that this process should be possible. Looking at the links/examples you have suplied I have created an Action which I think should do the job locally:</p>
<p><a href="https://github.com/jpadfield/simple-site/blob/master/.github/workflows/build.yml" rel="nofollow noopener">https://github.com/jpadfield/simple-site/blob/master/.github/workflows/build.yml</a></p>
<p>But it seems to be failing with a very unhelpfull error:</p>
<p>.github#L1<br>
Process completed with exit code 1.</p>
<p>Am I missing something simple?</p>
<p>Another related quesation, is it possible to limit the action to only run after certain files, or perhaps a certain folder is changed? Not a big issue, but it would stop the action running when it is not needed.</p>
<p>Thanks</p>
<p>Joe</p>

**Comment by vidavidorra** [comment, marked as answer]
<p>Looking at the logs it looks like the code fialed becuase there was noting to commit from the <code>Commit files</code> section.</p>
<pre><code>On branch master
Your branch is up to date with 'origin/master'.

nothing to commit, working tree clean
##[error]Process completed with exit code 1.
</code></pre>
<p>Meaning that the folder <code>./docs</code> doesn’t seem to exist at that location. I’m not familiar with PHP at all but could it be that the <code>php build.php</code> creates the docs (I assume) in <code>build/docs</code>?<br>
What happens if you run that command locally? Or add <code>ls -la ./build</code> or something like that to the workflow to debug.</p>

**Comment by jpadfield** [comment]
<p>Thank you :slight_smile:</p>
<p>So basically it was working, but because no updates where required it returned an error.</p>
<p>Ok no to find an if the files are different then commit them.</p>
<p>Joe</p>

**Reply by jairamjidgekar** [reply]
"Ok no to find an if the files are different then commit them." 

-- where did you make the changes to achieve the above? 
in GH Actions .yml file or  your programming script?

**Comment by jpadfield** [comment]
<p>Thank you all the Action is now working :slight_smile:</p>
<p>Joe</p>
<p>Just for reference here is the working file:</p>
<pre><code># This is a basic workflow to help you get started with Actions

name: Build GH Pages

# Controls when the action will run. Triggers the workflow on push or pull request
# events but only for the master branch
on:
  push:
    branches: ['master']
    paths: 'build/*.json'

# A workflow run is made up of one or more jobs that can run sequentially or in parallel
jobs:
  # This workflow contains a single job called "build"
  build:
    # The type of runner that the job will run on
    runs-on: ubuntu-latest

    # Steps represent a sequence of tasks that will be executed as part of the job
    steps:
    # Checks-out your repository under $GITHUB_WORKSPACE, so your job can access it
    - uses: actions/checkout@v2

    # Runs commands using the runners shell
    - name: Run the build php script
      run: |
        cd build
        php build.php # This file reads a few json files and then creates a set of html files in the doc folder
        cd ../
    - name: Commit files # transfer the new html files back into the repository
      run: |
        git config --local user.name "jpadfield"
        git add ./docs
        git commit -m "Updating the repository GitHub html pages in the docs folder"
    - name: Push changes # push the output folder to your repo
      uses: ad-m/github-push-action@master
      with:
        github_token: ${{ secrets.GITHUB_TOKEN }}
        force: true
</code></pre>

**Reply by jairamjidgekar** [reply]
Hi Joe, 

I am stuck with the similar problem. 

I have a .py script, which deletes an existing file and replaces with the new file (in the same path). I have scheduled this using GH Actions on a daily basis. 

On any given day, there might be chances that, there won't be any changes in the file downloaded. So, in that situation the GH Actions workflow will fail in the commit task, since there's no change. 

------------------------------------------------------------------------------------------------------------
Same error as you have mentioned in the above message:
------------------------------------------------------------------------------------------------------------
```
Your branch is up to date with 'origin/master'.
nothing to commit, working tree clean
##[error]Process completed with exit code 1.
```
------------------------------------------------------------------------------------------------------------


As a work around, what I did is separated the code (delete files and add files) into two different .py scripts and added commit tasks respectively. 

<img width="325" alt="image" src="https://user-images.githubusercontent.com/108509719/199168857-9540d37e-77d7-4b5e-8791-63826ba26325.png">

```
 - id: python_remove
        name: Python Remove file execution
        run: python Images/Remove_Files.py  
          # python Images/Images_of_ILO_Biography.py
      
      # Remove old files in folder,  push
      - id: remove_commit
        name: Commit files - Removes
        run:
          git config --local user.name actions-user

          git config --local user.email "actions@github.com" 
          
          git fetch

          git add *

          git commit -am "GH Action   File removed    $(date)" 
 
          git push -f origin main


 - id: python 
        name: Python file execution 
        run: python Images/Images_of_ILO_Biography.py  
          # python Images/Images_of_ILO_Biography.py
      
      # Add files in folder,  push
      - id: commit_files
        name: Commit files  
        run:
          git config --local user.name actions-user

          git config --local user.email "actions@github.com" 
          
          git fetch

          git add *

          git commit -am "GH Action   Files added    $(date)" 
 
          git push -f origin main
```

But this code seems to be recurring (I feel there should be some other option, if any) 

Let me know your views on this. thank you.

**Reply by jpadfield** [reply]
Hi,

This was a while ago, but I seem to remember that one of the ways I got round the no change needed error was to ensure that there always was a change. I added an extra file creation and deletion step that created a dummy file with the current date in the file name - you would need to pick a file extension that is already being deleted or a non standard one that will not effect what you are trying to do. This ensured that my action always did something and did not return errors (unless something else went wrong :-) ).

If you have a recurring issue it might be good to look at what actions trigger changes, one of the steps may be just re-triggering the whole process - so it just keep on going.

Joe

**Reply by airtower-luna** [reply]
Another option is to check (e.g. with [`git diff --exit-code`](https://git-scm.com/docs/git-diff#Documentation/git-diff.txt---exit-code)) if there actually are any changes, and skip the commit if not. Definitely better than committing deletion first and then (possibly identical) new files. Example:

```yaml
      - id: commit_files
        name: Commit files  
        run:
          git config --local user.name actions-user
          git config --local user.email "actions@github.com"
          if ! git diff --exit-code; then
              git add your/files/here
              git commit -am "GH Action Files added $(date)"
              git push -f origin main
          fi
```

However, if you're building a Pages site there's now a better option that didn't exist back in 2020: directly deploy the build results instead of committing them, see [Publishing with a custom GitHub Actions workflow](https://docs.github.com/en/pages/getting-started-with-github-pages/configuring-a-publishing-source-for-your-github-pages-site#publishing-with-a-custom-github-actions-workflow). :wink:

**Comment by vidavidorra** [comment]
<blockquote>
<p>So basically it was working, but because no updates where required it returned an error.<br>
Indeed </p>
</blockquote>
<p>Great, happy to help! </p>
<p>For future reference: by clicking on the action (in the left menu) you could see more of the logs than e.g. the following.</p>
<pre><code>.github#L1
Process completed with exit code 1.
</code></pre>

**Reply by jairamjidgekar** [reply]
@vidavidorra
Please have a look on my above comment/reply. Can you please help me.

**Comment by jpadfield** [comment]
<p><a class="mention" href="/vidavidorra">@vidavidorra</a> yes I had missed that one, so was working a bit blind, but once you mentioned it I explored some more and found it, so thanks.</p>
<p>Joe</p>

**Comment by mrchrisster** [comment]
<p>Just wanted to say thank you, this worked great for me!</p>

**Comment by sowmyareddyu** [comment]
Hi @jpadfield  I am trying to append file through python and run the workflow through self hosted runner. I get the following error with the below code. Could you please help me with it?

Run git config user.name github-actions
  git config user.name github-actions
  git config user.email github-actions@github.com
  git add .
  git commit -m "generated"
  git push


  shell: C:\Windows\System3[2](https://github.com/repo/actions/runs/4557587906/jobs/8039416704#step:6:2)\WindowsPowerShell\v1.0\powershell.EXE -command ". '{0}'"

    pythonLocation: C:\actions-runner\_work\_tool\Python\3.11.1\x64
    PKG_CONFIG_PATH: C:\actions-runner\_work\_tool\Python\3.11.1\x64/lib/pkgconfig
    Python_ROOT_DIR: C:\actions-runner\_work\_tool\Python\3.11.1\x64
    Python2_ROOT_DIR: C:\actions-runner\_work\_tool\Python\3.11.1\x64
    Python3_ROOT_DIR: C:\actions-runner\_work\_tool\Python\3.11.1\x64
**git : The term 'git' is not recognized as the name of a cmdlet, function, script file, or operable program. Check the 
spelling of the name, or if a path was included, verify that the path is correct and try again.**
At C:\actions-runner\_work\_temp\dc1bbb01-6c36-4b4e-be47-649c715a826e.ps1:2 char:1
+ git config user.name github-actions
+ ~~~
    + CategoryInfo          : ObjectNotFound: (git:String) [], ParentContainsErrorRecordException
    + FullyQualifiedErrorId : CommandNotFoundException
 
Error: Process completed with exit code 1.

**Reply by jpadfield** [reply]
Sorry not sure I can help with this one.

**Reply by airtower-luna** [reply]
The error message sounds like Git isn't actually installed on your runner.

**Reply by sowmyareddyu** [reply]
@airtower-luna right, that was the problem. I installed git in runner and ran git init. I am also able to do git --version. but when I run the above code, it fails with git is not recognised.

![image](https://user-images.githubusercontent.com/27121128/229101018-29c684bb-e325-4473-bf0f-dc6f03383d58.png)

**Reply by airtower-luna** [reply]
In the step you posted you *manually* add the Git install location to the PATH. You'll either make sure Git is in the default path (I'm not sure how to do that on Windows, but that's the thing to look for), or do that in every job on your runner (see [Adding a system path](https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#adding-a-system-path) for a per-job way to do that).

**Reply by Encephala** [reply]
Note every `run` in a workflow executes in a separate shell and thus has its environment reset again. You have to also add the line to expand `PATH` to the job which executes the git commands

**Comment by Encephala** [comment]
I'd like to chime on how I figured out how to push without depending on an external action. 
I wanted to use the python package [`coverage-badge`](https://pypi.org/project/coverage-badge/) to automatically generate a badge reporting the coverage (in a Django app) to put in the README, below is the workflow YAML.

GitHub by default adds `GITHUB_TOKEN` to your secrets on every action. This token doesn't have write permissions to the repository by default, so you have to add the `permissions: contents: write`-part to allow it to make changes.
Then, as @logankilpatrick showed above (🙏) you can make git authenticate with this token by changing the remote URL.

```yml
jobs:
  build:
    runs-on: ubuntu
    permissions:
      contents: write
    steps:
      - uses: actions/checkout@v4

     [...]

      - name: Generate coverage badge
        working-directory: ./app
        run: |
          coverage run --source='.' --omit="*/migrations/*,app/*,manage.py" manage.py test
          coverage-badge -o ../coverage.svg -f
          git config --local user.name "github-actions[bot]"
          git config --local user.email "github-actions[bot]@users.noreply.github.com"
          git add ../coverage.svg
          if [ -z "$(git status --porcelain)" ]; then
            echo "No changes to commit"
            exit 0
          fi
          git commit -m "<commit message>"
          git remote set-url origin https://<username>:${{ secrets.GITHUB_TOKEN }}@github.com/${{ github.repository }}
          git push origin main
```
(The `if` statement ensures the action doesn't fail if the coverage percentage didn't update)

**Reply by guest271314** [reply]
Is there a way to run the actions on each request of a GitHub Pages request?
